File: src/main/java/org/cerion/stocklist/YahooFinance.java
package org.cerion.stocklist;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.MalformedURLException;
import java.net.URL;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

public class YahooFinance 
{

	public static final int INTERVAL_DAILY = 0;
	public static final int INTERVAL_WEEKLY = 1;
	public static final int INTERVAL_MONTHLY = 2;

	protected static DateFormat mDateFormat = new SimpleDateFormat("yyyy-MM-dd");

	public static PriceList getPrices(String sSymbol, int interval, int count)
	{
		Calendar cal = Calendar.getInstance();

		if(interval == INTERVAL_DAILY) //This one is just an estimate since there are various days the market is closed
		{
			final int trading_days_year = 252;
			int remaining = count;
			while(remaining >= trading_days_year) 
			{
				cal.add(Calendar.YEAR, -1);
				remaining -= trading_days_year;
			}
			
			int weekdays = (int) (count * (365.0 / trading_days_year ));
			cal.add(Calendar.DAY_OF_YEAR, 1-weekdays);
		}
		else if(interval == INTERVAL_WEEKLY)
		{
			cal.set(Calendar.DAY_OF_WEEK, Calendar.SUNDAY);
			cal.add(Calendar.WEEK_OF_MONTH, (1-count)); 
		}
		else if(interval == INTERVAL_MONTHLY)
		{
			cal.add(Calendar.MONTH, (1-count));
			cal.set(Calendar.DAY_OF_MONTH, 1);
		}
		
		return getPrices(sSymbol,interval,cal);
	}
	
	public static PriceList getPrices(String sSymbol, int interval, Calendar calendar)
	{
		String sURL = "http://ichart.yahoo.com/table.csv?s=" + sSymbol;
		Calendar cal = calendar;
		if(interval == INTERVAL_MONTHLY)
		{
			sURL += "&g=m"; //monthly (d/w/m and v=dividends)
			if(cal == null) 
				cal = new GregorianCalendar(1970,0,1);
		}
		else if(interval == INTERVAL_WEEKLY)
		{
			sURL += "&g=w"; //weekly
			if(cal == null)
				cal = new GregorianCalendar(2000,0,1);
		}
		else if(cal == null)
			cal = new GregorianCalendar(2000,0,1); //October 9, 2005
		
		//a = month minus 1
		//b = day of month
		//c = year
		sURL += String.format("&a=%d&b=%d&c=%d", cal.get(Calendar.MONTH), cal.get(Calendar.DAY_OF_MONTH), cal.get(Calendar.YEAR));
		
		System.out.println(sURL);
		
		String sData = getURL(sURL);
		
		String lines[] = sData.split("\\r\\n");
		

		ArrayList<Price> prices = new ArrayList<>();
		for(int i = 1; i < lines.length; i++)
			prices.add(parseLine(lines[i]));
		
		PriceList result = new PriceList(sSymbol,prices);
		
		System.out.println(result.size() + " " + sURL);
		
		return result;
	}
	

	private static Price parseLine(String sLine)
	{
		String fields[] = sLine.split(",");
		if(fields.length == 7)
		{
			//TODO fix this for S&P large numbers
			try //Fails on S&P index since too large, just ignore it
			{
				float open = Float.parseFloat(fields[1]);
				float high = Float.parseFloat(fields[2]);
				float low  = Float.parseFloat(fields[3]);
				float close = Float.parseFloat(fields[4]);
				float adjClose = Float.parseFloat(fields[6]);
				
				if(adjClose != close)
				{
					open = (adjClose * open) / close;
	
					if(close == high) //Fix float rounding issues to prevent close > high when they are the same
						high = adjClose;
					else
						high = (adjClose * high) / close;
					
					if(close == low)
						low = adjClose;
					else
						low = (adjClose * low) / close;
				}
				
				long volume = Long.parseLong(fields[5], 10);
				volume /= 1000;
				Date date = mDateFormat.parse( fields[0] );
			
				return new Price(date, open, high, low, adjClose, volume);
			}
			catch(Exception e)
			{		
				e.printStackTrace();
			}
		}
		
		return null;
	}
	
	
	public static ArrayList<Dividend> getDividends(String symbol)
	{
		ArrayList<Dividend> result = new ArrayList<>();
		String sURL = "http://ichart.yahoo.com/table.csv?s=" + symbol + "&g=v";
		sURL += "&a=1&b=1&c=2000";
		
		String sData = getURL(sURL);
		String lines[] = sData.split("\\r\\n");
		for(int i = 1; i < lines.length; i++)
		{
			String line[] = lines[i].split(",");
			
			Dividend div = new Dividend(parseDate(line[0]), Float.parseFloat(line[1]) );
			result.add(0, div);
		}
		
		
		System.out.println(symbol + "\t" + result.size());
		return result;
	}
	
	public static Date parseDate(String date)
	{
		Date result = new Date();
		try 
		{
			result = mDateFormat.parse( date );
		}
		catch(Exception e)
		{		
			System.out.println(e);
		}
		
		return result;
	}
	
	//TODO, make protected then maybe private
    public static String getURL(String theUrl)
    {
    	String sResult = null;
    	
    	//TODO try with resources
        try 
        {
            URL gotoUrl = new URL(theUrl);
            InputStreamReader isr = new InputStreamReader(gotoUrl.openStream());
            BufferedReader in = new BufferedReader(isr);
            StringBuffer sb = new StringBuffer();
            String inputLine;
            //grab the contents at the URL
            while ((inputLine = in.readLine()) != null)
                sb.append(inputLine+"\r\n");
            
            sResult = sb.toString();
        }
        catch (MalformedURLException mue) 
        {
            mue.printStackTrace();
        }
		catch (IOException e) 
		{
			sResult = "";
		}
		
        return sResult;
    }
}

File: src/jvmMain/kotlin/web/clients/YahooFinance.kt
package org.cerion.stocks.core.web.clients

import org.cerion.stocks.core.PriceRow
import org.cerion.stocks.core.model.Dividend
import org.cerion.stocks.core.platform.KMPDate
import org.cerion.stocks.core.web.FetchInterval
import org.cerion.stocks.core.web.PriceHistoryDataSource
import org.cerion.stocks.core.web.Tools
import java.text.SimpleDateFormat
import java.util.*

class YahooFinance private constructor() : PriceHistoryDataSource {

    // TODO add interval enum specific to this class for more type safety, the other has values this API does not support

    private var mCookieCrumb: String? = null
    private var mCookie: String? = null

    override fun getPrices(symbol: String, interval: FetchInterval, start: Date?): List<PriceRow> {
        if (!setCookieCrumb())
            throw RuntimeException("Failed to get cookie")

        //https://query1.finance.yahoo.com/v7/finance/download/OHI
        // ?period1=1493915553
        // &period2=1496593953
        // &interval=1d
        // &events=history
        // &crumb=TSV3DSdPIjI

        var sURL = "https://query1.finance.yahoo.com/v7/finance/download/$symbol"
        if (start != null)
            sURL += "?period1=" + start.time / 1000
        else
            sURL += "?period1=-1325635200" // This is the date they use for S&P 500 index at max size

        sURL += "&period2=" + Date().time / 1000

        if (interval === FetchInterval.MONTHLY)
            sURL += "&interval=1mo"
        else if (interval === FetchInterval.WEEKLY)
            sURL += "&interval=1wk"
        else
            sURL += "&interval=1d"

        sURL += "&events=history"
        sURL += "&crumb=" + mCookieCrumb!!

        println(sURL)
        val res = Tools.getURL(sURL, mCookie)
        if (DEBUG) {
            println("Response size = " + res.result.length)
            if (res.result.length < 1000)
                println(res.result)
        }

        val sData = res.result
        return getPricesFromTable(sData)
    }

    fun getDividends(symbol: String): List<Dividend> {
        if (!setCookieCrumb())
            throw RuntimeException("Failed to get cookie")

        var sURL = "https://query1.finance.yahoo.com/v7/finance/download/$symbol"
        sURL += "?period1=946684800" // Jan 1, 2000
        sURL += "&period2=" + Date().time / 1000
        sURL += "&interval=1wk&events=div"
        sURL += "&crumb=" + mCookieCrumb!!

        val result = ArrayList<Dividend>()
        val res = Tools.getURL(sURL, mCookie)

        val sData = res.result
        val lines = sData.split("\\r\\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (i in 1 until lines.size) {
            val line = lines[i].split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

            val div = Dividend(parseDate(line[0])!!, java.lang.Float.parseFloat(line[1]))
            result.add(0, div)
        }

        return result
    }

    private fun setCookieCrumb(): Boolean {
        if (mCookieCrumb != null)
            return true

        val res = Tools.getURL("https://finance.yahoo.com/quote/%5EGSPC/options", null) // most any page will work here
        val page = res.result
        var index = page.indexOf("\"CrumbStore\"")
        if (index > 0) {
            if (DEBUG) {
                val debug = page.substring(index, index + 50)
                println(debug)
            }

            index = page.indexOf("\"crumb\"", index)

            if (index > 0) {
                val start = page.indexOf("\"", index + 8) + 1
                val end = page.indexOf("\"", start)
                if (start < end) {
                    mCookieCrumb = page.substring(start, end)
                    mCookieCrumb = mCookieCrumb!!.replace("\\u002F", "/")

                    // Seems to be different for local vs android, if more than 1 get the last entry
                    // If this still fails look into better method, might be difference between http vs https requests
                    val cookieHeaders = res.headers["Set-Cookie"]!!
                    mCookie = cookieHeaders[cookieHeaders.size - 1]
                    mCookie = mCookie!!.split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]
                    return true
                }
            }
        }

        return false
    }

    companion object {

        private val mDateFormat = SimpleDateFormat("yyyy-MM-dd")
        private val mDateFormatAlt = SimpleDateFormat("M/dd/yyyy")

        val instance = YahooFinance()

        private const val DEBUG = true

        /***
         * Gets PriceList from csv formatted file that API would return
         * @param tableData file contents as string
         * @return PriceList
         */
        fun getPricesFromTable(tableData: String): MutableList<PriceRow> {
            val lines = tableData.split("\\r\\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

            if (DEBUG)
                println("Table lines = " + lines.size)

            val prices = ArrayList<PriceRow>()
            for (i in 1 until lines.size) {
                if (!lines[i].contains("null"))
                    prices.add(parseLine(lines[i]))
                else
                    println("Ignoring line " + lines[i]) // new API issue some rows have all null values
            }

            return prices
        }

        private fun parseDate(inputDate: String): KMPDate? {
            var date = inputDate
            var result: Date?
            date = date.replace("\"", "")
            try {
                result = mDateFormat.parse(date)
            } catch (e: Exception) {
                // Try alternate
                try {
                    result = mDateFormatAlt.parse(date)
                } catch (ee: Exception) {
                    result = null
                    //ee.printStackTrace();
                }

            }

            return if (result != null)
                KMPDate(result)
            else
                null
        }

        fun parseLine(sLine: String): PriceRow {
            val fields = sLine.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (fields.size == 7) {
                //TODO fix this for S&P large numbers
                try
                //Fails on S&P index since too large, just ignore it
                {
                    var open = java.lang.Float.parseFloat(fields[1])
                    var high = java.lang.Float.parseFloat(fields[2])
                    var low = java.lang.Float.parseFloat(fields[3])
                    val close = java.lang.Float.parseFloat(fields[4])
                    val adjClose = java.lang.Float.parseFloat(fields[5])

                    if (adjClose != close) {
                        if (close == open)
                            open = adjClose
                        else
                            open = adjClose * open / close

                        if (close == high)
                        //Fix float rounding issues to prevent close > high when they are the same
                            high = adjClose
                        else
                            high = adjClose * high / close

                        if (close == low)
                            low = adjClose
                        else
                            low = adjClose * low / close
                    }

                    // Correcting bad data
                    if (open < low) {
                        println("Correcting bad [open] data on " + fields[0])
                        open = (high + low) / 2
                    }

                    var volume = java.lang.Long.parseLong(fields[6], 10)
                    volume /= 1000
                    val date = KMPDate.parse(fields[0])

                    return PriceRow(date, open, high, low, adjClose, volume.toFloat())
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }

            throw Exception("Unexpected price line")
        }
    }
}
